Fix bug when doing linked schema change.

chaoyli opened the pull request:
1. rollup will create table with num_key_columns is less than base table.
2. adding column to existed table, num_key_columns is larger than base table.
When meets two events, we should handle them carefully.

Commit:
Fix bug when doing linked schema change.

1. rollup will create table with num_key_columns is less than base table.
2. adding column to existed table, num_key_columns is larger than base table.
When meets two events, we should handle them carefully.

imay commented on be/src/olap/rowset/alpha_rowset_writer.cpp:
```diff
@@ -129,7 +129,28 @@ OLAPStatus AlphaRowsetWriter::add_rowset(RowsetSharedPtr rowset) {
+    // this is feasible because LinkedSchemaChange is done on the same disk
+    AlphaRowsetSharedPtr alpha_rowset = std::dynamic_pointer_cast<AlphaRowset>(rowset);
+    for (auto& segment_group : alpha_rowset->_segment_groups) {
+        RETURN_NOT_OK(_init());
```
_init will be called many times

imay commented on be/src/olap/rowset/segment_group.cpp:
```diff
@@ -234,27 +235,42 @@ bool SegmentGroup::delete_all_files() {
         DCHECK(second != NULL) << "failed to allocate memory for field: " << i;
-        second->copy(zone_map_fields[i].second);
+
+        //for new key column, use default value to fill into column_statistics
```
```suggestion
        // for new key column, use default value to fill into column_statistics
```

imay commented on be/src/olap/rowset/segment_group.cpp:
```diff
@@ -234,27 +235,42 @@ bool SegmentGroup::delete_all_files() {
-    //Should use _num_key_columns, not zone_map_fields.size()
-    //as rollup tablet num_key_columns will less than base tablet zone_map_fields.size().
-    //For LinkedSchemaChange, the rollup tablet keys order is the same as base tablet
+    //1. rollup tablet num_key_columns() will less than base tablet zone_map_fields.size().
```
```suggestion
    // 1. rollup tablet num_key_columns() will less than base tablet zone_map_fields.size().
```

imay commented on be/src/olap/rowset/segment_group.cpp:
```diff
@@ -234,27 +235,42 @@ bool SegmentGroup::delete_all_files() {
-    //For LinkedSchemaChange, the rollup tablet keys order is the same as base tablet
+    //1. rollup tablet num_key_columns() will less than base tablet zone_map_fields.size().
+    //   For LinkedSchemaChange, the rollup tablet keys order is the same as base tablet
+    //2. adding column to existed table, num_key_columns() will larger than
```
```suggestion
    // 2. adding column to existed table, num_key_columns() will larger than
```

Commit:
Fix bug when doing linked schema change.

1. rollup will create table with num_key_columns is less than base table.
2. adding column to existed table, num_key_columns is larger than base table.
   When meets two events, we should handle them carefully.

imay reviewed (approved):
LGTM

yiguolei reviewed (approved):
LGTM
